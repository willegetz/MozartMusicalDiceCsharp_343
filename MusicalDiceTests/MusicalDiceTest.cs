using System;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using System.Text;
using System.IO;
using MozartMusicalDice;
using ApprovalTests.Reporters;
using ApprovalTests;

namespace MusicalDiceTests
{
    [TestClass]
    [UseReporter(typeof(BeyondCompareReporter))]
    public class MusicalDiceTest
    {
        // A measure is 3 beats and is 0 based. Measure 1 would be beat 0, 1, and 2 but not 3. Measure 144 would be beat 432, 422, and 434 but not 435.
        // Given a set of measures, extract the appropriate beats for a given measure.

        private static MusicalDice musicDice;

        [ClassInitialize]
        public static void LoadSampleComposition(TestContext testContext)
        {
            musicDice = new MusicalDice();
        }

        [TestMethod]
        public void Measure1HasBeats0UpTo3()
        {
            var beatsForMeasure1 = musicDice.RetrieveBeatsForMeasure(1);

            Approvals.Verify(beatsForMeasure1);
        }

        [TestMethod]
        public void Measure22HasBeats63UpTo66()
        {
            var beatsForMeasure22 = musicDice.RetrieveBeatsForMeasure(22);

            Approvals.Verify(beatsForMeasure22);
        }

        [TestMethod]
        public void Measure176HasBeat525UpTo528(){
            var beatsForMeasure176 = musicDice.RetrieveBeatsForMeasure(176);

            Approvals.Verify(beatsForMeasure176);
        }

        [TestMethod]
        public void GenerateRawCompositionFrom22nd176thAnd1stMeasures()
        {
            var composition = new StringBuilder();

            var firstMeasure = musicDice.RetrieveBeatsForMeasure(22);
            composition.Append(firstMeasure).Append("\n");

            var secondMeasure = musicDice.RetrieveBeatsForMeasure(176);
            composition.Append(secondMeasure).Append("\n");

            var thirdMeasure = musicDice.RetrieveBeatsForMeasure(1);
            composition.Append(thirdMeasure);

            var rawComposition = composition.ToString().Trim();
            
            Approvals.Verify(rawComposition);
        }

        // Measure 22 (beats 63 to 66) to become Measure 1 (beats 0 to 3)
        [TestMethod]
        public void Measure22ToBecomeMeasure1()
        {
            var currentMeasure = 22;
            var newMeasure = 1;

            var beatOffset = musicDice.GetBeatOffsetForMeasures(currentMeasure, newMeasure);

            var expected = -63;

            Assert.AreEqual(expected, beatOffset);
        }

        // Measure 176 (beats 525 to 528) to become Measure 6 (beats 15 to 18)
        [TestMethod]
        public void Measure176ToBecomeMeasure6()
        {
            var currentMeasure = 176;
            var newMeasure = 6;

            var beatOffset = musicDice.GetBeatOffsetForMeasures(currentMeasure, newMeasure);

            var expected = -510;

            Assert.AreEqual(beatOffset, expected);
        }

        // Measure 8 (21 to 24) to become Measure 13 (36 to 39)
        [TestMethod]
        public void Measure8ToBecomeMeasure13()
        {
            var currentMeasure = 8;
            var newMeasure = 13;

            var beatOffset = musicDice.GetBeatOffsetForMeasures(currentMeasure, newMeasure);

            var expected = 15;

            Assert.AreEqual(beatOffset, expected);
        }

        [TestMethod]
        public void Measure22BeatsCalculatedToBeMeasure1Beats()
        {
            var currentMeasure = 22;
            var newMeasure = 1;

            var currentMeasureBeats = musicDice.RetrieveBeatsForMeasure(currentMeasure);
            var beatOffset = musicDice.GetBeatOffsetForMeasures(currentMeasure, newMeasure);
            var newMeasureBeats = musicDice.GetAdjustedBeatsForNewMeasure(currentMeasureBeats, beatOffset);

            Approvals.Verify(newMeasureBeats);
        }

        [TestMethod]
        public void Measure176BeatsCalculatedToBeMeasure6Beats()
        {
            var currentMeasure = 176;
            var newMeasure = 6;

            var currentMeasureBeats = musicDice.RetrieveBeatsForMeasure(currentMeasure);
            var beatOffset = musicDice.GetBeatOffsetForMeasures(currentMeasure, newMeasure);
            var newMeasureBeats = musicDice.GetAdjustedBeatsForNewMeasure(currentMeasureBeats, beatOffset);

            Approvals.Verify(newMeasureBeats);
        }

        [TestMethod]
        public void Measure8BeatsCalculatedToBeMeasure13Beats()
        {
            var currentMeasure = 8;
            var newMeasure = 13;

            var currentMeasureBeats = musicDice.RetrieveBeatsForMeasure(currentMeasure);
            var beatOffset = musicDice.GetBeatOffsetForMeasures(currentMeasure, newMeasure);
            var newMeasureBeats = musicDice.GetAdjustedBeatsForNewMeasure(currentMeasureBeats, beatOffset);

            Approvals.Verify(newMeasureBeats);
        }

        [TestMethod]
        public void GenerateNewCompositionFrom22nd176thAnd1stMeasures()
        {
            var composition = new StringBuilder();

            var measure22 = 22;
            var measure176 = 176;
            var measure1 = 1;

            var firstMeasure = 1;
            var secondMeasure = 2;
            var thirdMeasure = 3;

            var beatsForMeasure22 = musicDice.RetrieveBeatsForMeasure(measure22);
            var firstMeasureBeatOffset = musicDice.GetBeatOffsetForMeasures(measure22, firstMeasure);
            var newFirstMeasure = musicDice.GetAdjustedBeatsForNewMeasure(beatsForMeasure22, firstMeasureBeatOffset);
            composition.Append(newFirstMeasure).Append("\n");

            var beatsForMeasure176 = musicDice.RetrieveBeatsForMeasure(measure176);
            var secondMeasureBeatOffset = musicDice.GetBeatOffsetForMeasures(measure176, secondMeasure);
            var newSecondMeasure = musicDice.GetAdjustedBeatsForNewMeasure(beatsForMeasure176, secondMeasureBeatOffset);
            composition.Append(newSecondMeasure).Append("\n");

            var beatsForMeasure1 = musicDice.RetrieveBeatsForMeasure(measure1);
            var thirdMeasureBeatOffset = musicDice.GetBeatOffsetForMeasures(measure1, thirdMeasure);
            var newThirdMeasure = musicDice.GetAdjustedBeatsForNewMeasure(beatsForMeasure1, thirdMeasureBeatOffset);
            composition.Append(newThirdMeasure).Append("\n");

            var newComposition = composition.ToString().Trim();

            Approvals.Verify(newComposition);
        }

        [TestMethod]
        public void GenerateNewCompositionFromArrayOfMeasures()
        {
            // create array of measures from sample matrix in challengeinfo.txt
            var measuresArray = new[] { 96, 22, 141, 41, 105, 122, 11, 30, 70, 121, 26, 9, 112, 49, 109, 14 };

            // pass array of measures into the generate new composition function on the musical dice object
            var newComposition = musicDice.CreateNewComposition(measuresArray);

            // verify the new composition is correct.
            Approvals.Verify(newComposition);
        }
    }
}
