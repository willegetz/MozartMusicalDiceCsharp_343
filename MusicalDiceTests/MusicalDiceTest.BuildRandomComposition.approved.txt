C3 0 2
E3 0 2
G5 0 1
C5 1 1
E5 2 1
C3 3 2
E3 3 2
G5 3 1
C5 4 1
C3 5 1
E5 5 1
G3 5 1
B2 6 2
D5 6 1
B4 7 1
G4 8 1
C5 9 1
E3 9 2
G3 9 2
E5 10 0.5
C5 10.5 0.5
G5 11 1
C3 12 1
D3 12 1
F#5 12 1
C3 13 1
D3 13 1
F#5 13 0.5
D5 13.5 0.5
A5 14 1
C3 14 1
D3 14 1
B2 15 2
D3 15 2
G5 15 1
G5 16 0.5
D5 16.5 0.5
B5 17 1
C3 18 1
E5 18 0.5
C6 18.5 0.5
B5 19 0.5
D3 19 1
G5 19.5 0.5
A5 20 0.5
D2 20 1
F#5 20.5 0.5
B4 21 2
D5 21 2
G2 21 1
G4 21 2
G5 21 2
G3 22 0.5
F3 22.5 0.5
E3 23 0.5
D3 23.5 0.5
A3 24 2
C3 24 2
F#5 24 0.5
D5 24.5 0.5
A5 25 1
A3 26 1
C3 26 1
F#5 26 1
G2 27 0.5
G5 27 1
B2 27.5 0.5
D5 28 2
G3 28 1
B2 29 1
C3 30 2
E5 30 0.5
G3 30 2
C5 30.5 0.5
G4 31 1
C3 32 1
E5 32 1
G3 32 1
D5 33 1
G3 33 0.5
F#3 33.5 0.5
G3 34 0.5
G4 34 2
D3 34.5 0.5
B2 35 0.5
G2 35.5 0.5
C3 36 0.5
E3 36 0.5
G5 36 1
G3 36.5 0.5
C3 37 0.5
C5 37 1
E3 37 0.5
G3 37.5 0.5
C3 38 0.5
E3 38 0.5
E5 38 1
G3 38.5 0.5
C3 39 0.5
E3 39 0.5
G5 39 1
G3 39.5 0.5
C3 40 0.5
C5 40 1
E3 40 0.5
G3 40.5 0.5
C3 41 0.5
E3 41 0.5
E5 41 1
G3 41.5 0.5
F3 42 2
F5 42 0.5
A5 42.5 0.5
A4 43 1
B4 44 0.5
G3 44 1
D5 44.5 0.5
C3 45 2
C5 45 1
C4 46 1
C2 47 1